edb/lang/graphql/types.py

```diff
@@ -153,36 +178,72 @@ def get_fields(self, typename):
                 ptr = edb_type.resolve_pointer(self.edb_schema, name)
                 target = self._get_target(ptr)
                 if target:
-                    args = (self.get_args(ptr.target.name)
-                            if isinstance(ptr.target, ObjectType) else None)
+                    args = None
+                    if isinstance(ptr.target, ObjectType):
+                        args = {
+                            'filter': GraphQLArgument(
+                                self._gql_inobjtypes[ptr.target.name]),
+                        }
                     fields[name.name] = GraphQLField(target, args=args)
 
         return fields
 
     @lru_cache(maxsize=None)
-    def get_args(self, typename):
-        args = OrderedDict()
+    def get_input_fields(self, typename):
+        selftype = self._gql_inobjtypes[typename]
+        fields = OrderedDict()
+        fields['and'] = GraphQLInputObjectField(
+            GraphQLList(GraphQLNonNull(selftype)))
+        fields['or'] = GraphQLInputObjectField(
+            GraphQLList(GraphQLNonNull(selftype)))
+        fields['not'] = GraphQLInputObjectField(selftype)
 
         edb_type = self.edb_schema.get(typename)
         for name in sorted(edb_type.pointers, key=lambda x: x.name):
             if name.name == '__type__':
                 continue
+            if name.name in fields:
+                raise GraphQLCoreError(
+                    f"{name.name!r} of {typename} clashes with special "
+                    "reserved fields required for GraphQL conversion"
+                )
 
             ptr = edb_type.resolve_pointer(self.edb_schema, name)
 
             if not isinstance(ptr.target, ScalarType):
                 continue
 
             target = self._convert_edb_type(ptr.target)
-            if target:
-                args[name.name] = GraphQLArgument(target)
+            intype = self._gql_inobjtypes.get(f'Filter{target.name}')
+            if intype:
+                fields[name.name] = GraphQLInputObjectField(intype)
 
-        return args
+        return fields
+
+    def define_generic_input_types(self):
+        eq = ['eq', 'neq']
+        comp = eq + ['gte', 'gt', 'lte', 'lt']
+        string = comp + ['like', 'ilike']
+
+        self._make_generic_input_type(GraphQLBoolean, *eq)
+        self._make_generic_input_type(GraphQLID, *eq)
+        self._make_generic_input_type(GraphQLInt, *comp)
+        self._make_generic_input_type(GraphQLFloat, *comp)
+        self._make_generic_input_type(GraphQLString, *string)
+
+    def _make_generic_input_type(self, base, *ops):
```

Comment:
No need to use `*args` here as far as I can see.

Reply:
Oh, that's right, this is an unnecessary complication at this point.